GitHub Actions: Support configuring triggers for a workflow

kevmoo opened the pull request:
Fixes https://github.com/google/mono_repo.dart/issues/239

Breaking changes to how CI is configured. See the changelog.

natebosch commented on mono_repo/lib/src/commands/travis.dart:
```diff
@@ -57,7 +57,7 @@ class CIConfigMismatchException extends UserException {
 /// Checks that the root CI config includes exactly one provider, travis.
 void _checkCIConfig(RootConfig rootConfig) {
-  if (rootConfig.monoConfig.ci != [CI.travis]) {
+  if (rootConfig.monoConfig.ci != {CI.travis}) {
```
Is this doing what you want it to? Should this be using `SetEquality` from `package:collection`?

kevmoo replied:
Yeah...I was wondering that when I refactored this...hrm...

kevmoo replied:
fixed

kevmoo replied:
right, @jakemac53 ?

natebosch commented on mono_repo/lib/src/github_config.dart:
```diff
@@ -0,0 +1,55 @@
+import 'package:json_annotation/json_annotation.dart';
```
copyright?

kevmoo replied:
fixed

kevmoo commented:
@jakemac53 @natebosch – if you like we can roll this out - with cron – and ditch travis is the places

jakemac53 commented on mono_repo/lib/src/github_config.dart:
```diff
@@ -0,0 +1,59 @@
+class GitHubConfig {
+  final Map<String, dynamic> on;
+
+  // This is currently just to make JsonSerializable happy – need to fix!
```
Add a TODO with a bug link please so this can be cleaned up more easily, or if you don't think it will be cleaned up then just mention it is a workaround.

jakemac53 commented on mono_repo/lib/src/github_config.dart:
```diff
@@ -0,0 +1,59 @@
+  'push': {
+    'branches': [r'$default-branch']
+  },
+  'pull_request': null,
```
It might be good to clarify what `null` means here - I assume it means pull requests to any branch?

kevmoo replied:
ack

jakemac53 commented on mono_repo/lib/src/mono_config.dart:
```diff
@@ -132,9 +137,18 @@ class MonoConfig {
     final travis = json['travis'] ?? {};
 
-    if (travis is! Map) {
+    if (travis is bool) {
```
Consider breaking this out into a function, the check is identical between the two CI configs (you would have to pass in the `ci` set so you can remove them, or add a `!= false` check above).

kevmoo replied:
ack

jakemac53 commented:
We also need to update the `README.md`, I added docs for the `ci` key there which should be updated with the new way of doing things.

kevmoo commented:
> We also need to update the `README.md`, I added docs for the `ci` key there which should be updated with the new way of doing things.

I updated the readme – and now validate the content of the readme – which is nice :-)
